Sources/NIOTransportServices/StateManagedNWConnectionChannel.swift

```diff
@@ -0,0 +1,617 @@
+            }
+        default:
+            try self.setChannelSpecificOption0(option: option, value: value)
+        }
+    }
+
+    public func getOption<Option: ChannelOption>(_ option: Option) -> EventLoopFuture<Option.Value> {
+        if self.eventLoop.inEventLoop {
+            return self.eventLoop.makeCompletedFuture(Result { try getOption0(option: option) })
+        } else {
+            return eventLoop.submit { try self.getOption0(option: option) }
+        }
+    }
+
+    func getOption0<Option: ChannelOption>(option: Option) throws -> Option.Value {
+        self.eventLoop.assertInEventLoop()
+
+        guard !self.closed else {
+            throw ChannelError.ioOnClosedChannel
+        }
+
+        switch option {
+        case _ as ChannelOptions.Types.AutoReadOption:
+            return self.options.autoRead as! Option.Value
+        case _ as ChannelOptions.Types.SocketOption:
+            let optionValue = option as! ChannelOptions.Types.SocketOption
+
+            // SO_REUSEADDR and SO_REUSEPORT are handled here.
+            switch (optionValue.level, optionValue.name) {
+            case (SOL_SOCKET, SO_REUSEADDR):
+                return Int32(self.reuseAddress ? 1 : 0) as! Option.Value
+            case (SOL_SOCKET, SO_REUSEPORT):
+                return Int32(self.reusePort ? 1 : 0) as! Option.Value
+            default:
+                return try self.nwOptions.valueFor(socketOption: optionValue) as! Option.Value
+            }
+        case _ as ChannelOptions.Types.WriteBufferWaterMarkOption:
+            return self._backpressureManager.waterMarks as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption:
+            return self.enablePeerToPeer as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSAllowLocalEndpointReuse:
+            return self.allowLocalEndpointReuse as! Option.Value
+        case _ as ChannelOptions.Types.AllowRemoteHalfClosureOption:
+            return self.options.supportRemoteHalfClosure as! Option.Value
+        case _ as NIOTSChannelOptions.Types.NIOTSWaitForActivityOption:
+            return self.options.waitForActivity as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSCurrentPathOption:
+            guard let currentPath = self.connection?.currentPath else {
+                throw NIOTSErrors.NoCurrentPath()
+            }
+            return currentPath as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSMetadataOption:
+            let optionValue = option as! NIOTSChannelOptions.Types.NIOTSMetadataOption
+            guard let connection = self.connection else {
+                throw NIOTSErrors.NoCurrentConnection()
+            }
+            return connection.metadata(definition: optionValue.definition) as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSMultipathOption:
+            return self.multipathServiceType as! Option.Value
```

Comment:
Why is this split, with the `StateManagedNWConnectionChannel` handling the `get` but not the `set`?

Reply:
That used to have a reason, but not anymore.